File: src/clang/magma/option/Option.c
#include "Option.h"
struct Option_R map(struct R(*mapper)(struct T));
struct T orElseGet(struct T(*other)());
struct Tuple_Boolean_T toTuple(struct T other);
struct void ifPresent(struct Consumer_T consumer);
struct T orElse(struct T other);
struct Option_T filter(struct Predicate_T predicate);
int isPresent();
struct R match(struct R(*ifPresent)(struct T), struct R(*ifEmpty)());
int isEmpty();
struct Option_T or(struct Option_T(*other)());
struct Option_R flatMap(struct Option_R(*mapper)(struct T));

File: src/java/magma/Option.java
package magma;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Stream;

interface Option<T> {
	<R> Option<R> map(Function<T, R> mapper);

	T orElse(T other);

	void ifPresent(Consumer<T> consumer);

	boolean isEmpty();

	Option<T> or(Supplier<Option<T>> other);

	T orElseGet(Supplier<T> other);

	<R> Option<R> flatMap(Function<T, Option<R>> mapper);

	Stream<T> stream();

	Tuple<Boolean, T> toTuple(T other);
}
